scopus-import/src/main/java/no/sikt/nva/scopus/conversion/files/ScopusFileConverter.java

```diff
@@ -99,30 +103,48 @@ public ScopusFileConverter(HttpClient httpClient, S3Client s3Client, Environment
     }
 
     public List<AssociatedArtifact> fetchAssociatedArtifacts(DocTp docTp) {
-        var associatedArtifactsFromXmlReferences = extractAssociatedArtifactsFromFileReference(docTp);
+        var associatedArtifactsFromXmlReferences = extractAssociatedArtifactsFromFileReference(docTp)
+                                                       .stream()
+                                                       .filter(File.class::isInstance)
+                                                       .filter(this::isValid)
+                                                       .toList();
         return !associatedArtifactsFromXmlReferences.isEmpty() ? associatedArtifactsFromXmlReferences
                    : extractAssociatedArtifactsFromDoi(docTp);
     }
 
+    private boolean isValid(AssociatedArtifact associatedArtifact) {
+        return nonNull(((File) associatedArtifact).getMimeType());
+    }
+
     private static CrossrefResponse toCrossrefResponse(String body) throws JsonProcessingException {
         return JsonUtils.dtoObjectMapper.readValue(body, CrossrefResponse.class);
     }
 
     private static String getFilename(HttpResponse<InputStream> response) {
-        return extractContentDisposition(response).map(ScopusFileConverter::extractFileNameFromContentDisposition)
-                   .orElseGet(() -> randomUUID() + FILE_NAME_DELIMITER + PDF_FILE_TYPE);
+        return response.headers().firstValue(Headers.CONTENT_DISPOSITION)
+                   .map(ScopusFileConverter::extractFileNameFromContentDisposition)
+                   .or(() -> extractFileNameFromUrl(response))
+                   .or(() -> randomFileNameWithFileTypeFromContentTypeHeader(response.headers()))
```

Comment:
Possibly tidy using *::extractFileNameFromUrl, Response::headers, *::randomFileName…

Reply:
It does not look like it is possible, can you come with example?

Reply:
Don't worry, my idea isn't a good one.